pennylane/devices/preprocess.py

```diff
@@ -178,29 +179,88 @@ def validate_device_wires(
+    #       should be one-shot and deferred. Every other method, if requiring no transforms
+    #       to be applied, is between the user and the capabilities of the device. The user
+    #       should be able to specify whatever they want as the MCM method as long as the
+    #       device supports it. Tree-traversal, being exclusive to default.qubit, should be
+    #       declared in the capabilities of the default.qubit device, then we will be able
+    #       to validate against it just like any other custom mcm method.
+    if mcm_method == "tree-traversal" and device.name != "default.qubit":
+        raise qml.QuantumFunctionError(
+            "The tree-traversal MCM method is only supported by the 'default.qubit' device."
+        )
+
+    if mcm_method is None or mcm_method == "deferred":
+        return  # no need to validate if requested method is deferred or if no method is requested.
+
+    if device.capabilities is None:
+        # If the device does not declare its supported mcm methods through capabilities,
+        # simply check that the requested mcm method is something we recognize.
+        if mcm_method not in ("deferred", "one-shot", "tree-traversal"):
+            raise qml.QuantumFunctionError(
+                f'Requested MCM method "{mcm_method}" unsupported by the device. Supported methods '
+                f'are: "deferred", "one-shot", and "tree-traversal".'
+            )
+        return
+
+    if mcm_method == "one-shot" and not shots_present:
+        raise qml.QuantumFunctionError(
+            'The "one-shot" MCM method is only supported with finite shots.'
+        )
+
+    if mcm_method not in device.capabilities.supported_mcm_methods:
+
+        supported_methods = device.capabilities.supported_mcm_methods + ["deferred"]
+        supported_method_strings = [f'"{m}"' for m in supported_methods]
+        raise qml.QuantumFunctionError(
+            f'Requested MCM method "{mcm_method}" unsupported by the device. Supported methods '
+            f"are: {', '.join(supported_method_strings)}."
+        )
+
+
 @transform
 def mid_circuit_measurements(
     tape: QuantumScript,
     device,
     mcm_config=MCMConfig(),
     **kwargs,  # pylint: disable=unused-argument
 ) -> tuple[QuantumScriptBatch, PostprocessingFn]:
-    """Provide the transform to handle mid-circuit measurements.
+    """Provide the transform to handle mid-circuit measurements."""
 
-    If the tape or device uses finite-shot, use the native implementation (i.e. no transform),
-    and use the ``qml.defer_measurements`` transform otherwise.
-    """
     if isinstance(mcm_config, dict):
         mcm_config = MCMConfig(**mcm_config)
     mcm_method = mcm_config.mcm_method
+    _validate_mcm_method(mcm_method, device, bool(tape.shots))
     if mcm_method is None:
-        mcm_method = "one-shot" if tape.shots else "deferred"
-
+        mcm_method = _default_mcm_method(device.capabilities, bool(tape.shots))
```

Comment:
These seem like things that belong in `setup_execution_config` instead.

Potentially out of scope for this PR,  but I'd rather not continue to have this function depend on the device.  Device instances should instead depend on this function.

 Do we even need this transform at all?

We could instead just do:
```
if execution_config.mcm_method == "one-shot":
    program.add_transform(qml.dynamic_one_shot, postselect_mode=execution_config.mcm_config.postselect_mode)
elif execution_config.mcm_method == "deferred":
    program.add_Transform(qml.defer_measurements)
```
when setting up the transform program.

Reply:
I guess it would make more sense. We could have a public facing function called `validate_mcm_method` that device developers can use in their `setup_execution_config`, then `mid_circuit_measurement` would just blindly dispatch to a concrete transform according to what's requested in the processed execution config.

Reply:
Although I would indeed consider this out of scope for this PR. If we're removing an argument from a public facing transform program we should do a deprecation cycle.

Reply:
So the question then is how public this one is?

Reply:
I would still want to have `mid_circuit_measurement` in the preprocess transform program, and have it handle dispatches to different things based on the requested error method. We can potentially move the validation and resolution of the `mcm_method` somewhere else, for example into the new `_resolve_execution_config` in `qnode.py`?

Reply:
In that case, I'll get rid of this PR and move the relevant logic to the default preprocessing.